source/navier_stokes.cc

```diff
@@ -1135,7 +1135,7 @@ NavierStokes<dim>::solve_nonlinear_system(const double initial_residual)
       std::cout.flags(flags);
     }
 
-  return step;
+  return {step + 1, n_tot_iterations};
```

Comment:
Since `step` is a counter, the actual number of nonlinear iterations is `step+1`.

Reply:
Is this always true? If we reach max steps (https://github.com/MeltPoolDG/adaflo/blob/7c60685ae3457c4ac37c26717191d7a9cccf4444/source/navier_stokes.cc#L833) that the increment is not needed is it?

Reply:
The problems are the break statements? Like https://github.com/MeltPoolDG/adaflo/blob/7c60685ae3457c4ac37c26717191d7a9cccf4444/source/navier_stokes.cc#L901

Reply:
> Is this always true? If we reach max steps (https://github.com/MeltPoolDG/adaflo/blob/7c60685ae3457c4ac37c26717191d7a9cccf4444/source/navier_stokes.cc#L833) that the increment is not needed is it?

In case the nonlinear solver does not converge after `max_nl_iterations` the return value for the number of nonlinear iterations would be `max_nl_iterations+1`. In MeltPoolDG if `nonlinear_iter>max_nl_iterations` it indicates that the Newton failed. I must admit that this is not very clear from the coding perspective, however for our use case it is sufficient. If you think it makes more sense, we could also correct the returned number of nonlinear iterations and extend the return value by a state `failure|success`. @peterrum what do you think?

Reply:
Personally, I would prefer to throw an exception in the case of no success to be consistent with deal.II. As it is now, the assert in MeltPoolDG (https://github.com/MeltPoolDG/MeltPoolDG-dev/blob/22a793c57b6334e9f2afe110a6ebc0a28c9e1152/source/flow/adaflo_wrapper.cpp#L333) does not work any more. `n_newton_steps == adaflo_params.max_nl_iteration` is true if the solver converged or not converged in the last step.

Reply:
An assert sounds reasonable. I'll introduce one asap.

Reply:
> As it is now, the assert in MeltPoolDG (https://github.com/MeltPoolDG/MeltPoolDG-dev/blob/22a793c57b6334e9f2afe110a6ebc0a28c9e1152/source/flow/adaflo_wrapper.cpp#L333) does not work any more. `n_newton_steps == adaflo_params.max_nl_iteration` is true if the solver converged or not converged in the last step.

That's why I changed it inside https://github.com/MeltPoolDG/MeltPoolDG-dev/pull/713/files.

Reply:
As far as I see it, https://github.com/MeltPoolDG/MeltPoolDG-dev/blob/f272237078647f2661867f4329c94cb9cc126cb8/source/flow/adaflo_wrapper.cpp#L344 will be always true?

Reply:
No, if the newton did not converge, `step==max_nl_iterations` and thus the returned value is `iter.first==max_nl_iterations+1`.

Reply:
Oh. I see. I think an exception would be cleaner. One can be defined easily like https://github.com/dealii/dealii/blob/a61eedda7cb9355ab00c31f66269d0447ccb3bec/include/deal.II/trilinos/nox.h#L33-L34.